File: student/cpp/decoder.cpp
#include <png.h>

#include <array>
#include <cstdint>
#include <exception>
#include <iostream>
#include <stdexcept>
#include <string>
#include <unordered_set>
#include <vector>

using Pixel = std::array<std::uint8_t, 3>;
using PixelGrid = std::vector<std::vector<Pixel>>;

/** Decode a row-major 2-D pixel array and return text without a newline. */
std::string decode(const std::string& codeType, const PixelGrid& pixels) {
    throw std::runtime_error("decoder not implemented for " + codeType);
}

PixelGrid loadPixels(const char* path) {
    png_image image{};
    image.version = PNG_IMAGE_VERSION;
    if (!png_image_begin_read_from_file(&image, path))
        throw std::runtime_error("cannot read PNG");
    image.format = PNG_FORMAT_RGB;
    std::vector<std::uint8_t> bytes(PNG_IMAGE_SIZE(image));
    if (!png_image_finish_read(&image, nullptr, bytes.data(), 0, nullptr))
        throw std::runtime_error("cannot decode PNG");

    PixelGrid pixels(image.height, std::vector<Pixel>(image.width));
    for (std::size_t y = 0; y < image.height; ++y)
        for (std::size_t x = 0; x < image.width; ++x) {
            const std::size_t offset = 3 * (y * image.width + x);
            pixels[y][x] = {bytes[offset], bytes[offset + 1], bytes[offset + 2]};
        }
    return pixels;
}

int main(int argc, char** argv) {
    static const std::unordered_set<std::string> supported = {
        "japanese_braille", "morse", "code39", "pzn", "ean13", "telepen",
        "dx_film_edge", "hccb", "aztec_rune", "code128",
        "data_matrix", "qr", "micro_qr", "rmqr", "aztec", "pdf417",
        "micro_pdf417", "databar_expanded_stacked", "maxicode"
    };
    if (argc != 3) {
        std::cerr << "usage: decoder CODE_TYPE IMAGE_PATH\n";
        return 2;
    }
    if (!supported.contains(argv[1])) {
        std::cerr << "unsupported code type: " << argv[1] << '\n';
        return 2;
    }
    try {
        std::cout << decode(argv[1], loadPixels(argv[2])) << '\n';
        return 0;
    } catch (const std::exception& error) {
        std::cerr << error.what() << '\n';
        return 1;
    }
}

File: student/java/Decoder.java
import java.awt.image.BufferedImage;
import java.io.File;
import java.util.Set;
import javax.imageio.ImageIO;

/** Starter implementation: compile with `javac Decoder.java`. */
public final class Decoder {
    public record Pixel(int red, int green, int blue) {}

    private static final Set<String> SUPPORTED_TYPES = Set.of(
        "japanese_braille", "morse", "code39", "pzn", "ean13", "telepen",
        "dx_film_edge", "hccb", "aztec_rune", "code128",
        "data_matrix", "qr", "micro_qr", "rmqr", "aztec", "pdf417",
        "micro_pdf417", "databar_expanded_stacked", "maxicode"
    );

    /** Decode a row-major 2-D pixel array and return text without a newline. */
    public static String decode(String codeType, Pixel[][] pixels) {
        throw new UnsupportedOperationException("decoder not implemented for " + codeType);
    }

    private static Pixel[][] loadPixels(String path) throws Exception {
        BufferedImage image = ImageIO.read(new File(path));
        if (image == null) throw new IllegalArgumentException("not a readable image: " + path);
        Pixel[][] pixels = new Pixel[image.getHeight()][image.getWidth()];
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                int rgb = image.getRGB(x, y);
                pixels[y][x] = new Pixel((rgb >>> 16) & 255, (rgb >>> 8) & 255, rgb & 255);
            }
        }
        return pixels;
    }

    public static void main(String[] args) {
        if (args.length != 2) {
            System.err.println("usage: java Decoder CODE_TYPE IMAGE_PATH");
            System.exit(2);
        }
        if (!SUPPORTED_TYPES.contains(args[0])) {
            System.err.println("unsupported code type: " + args[0]);
            System.exit(2);
        }
        try {
            System.out.println(decode(args[0], loadPixels(args[1])));
        } catch (Exception error) {
            System.err.println(error.getMessage());
            System.exit(1);
        }
    }
}
